pkg/encoding/kzgEncoder/multiframe.go

```diff
@@ -0,0 +1,171 @@
+	bls "github.com/Layr-Labs/eigenda/pkg/kzg/bn254"
+)
+
+// Sample is the basic unit for a verification
+// A blob may contain multiple Samples
+type Sample struct {
+	Commitment bls.G1Point
+	Proof      bls.G1Point
+	Row        int // corresponds to a row in the verification matrix
+	Coeffs     []bls.Fr
+	X          uint // X is assignment
+}
+
+// generate a random value using Fiat Shamir transform
+// we can also pseudo randomness generated locally, but we have to ensure no adv can manipulate it
+// Hashing everything takes about 1ms, so Fiat Shamir transform does not incur much cost
+func GenRandomness(params rs.EncodingParams, samples []Sample, m int) (bls.Fr, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+
+	for _, sample := range samples {
+		err := enc.Encode(sample.Commitment)
+		if err != nil {
+			return bls.ZERO, err
+		}
+	}
+
+	var randomFr bls.Fr
+
+	err := bls.HashToSingleField(&randomFr, buffer.Bytes())
+	if err != nil {
+		return bls.ZERO, err
+	}
+
+	return randomFr, nil
+}
+
+// UniversalVerify implements batch verification on a set of chunks given the same chunk dimension (chunkLen, numChunk).
+// The details is given in Ethereum Research post whose authors are George Kadianakis, George Kadianakis, George Kadianakis
+// https://ethresear.ch/t/a-universal-verification-equation-for-data-availability-sampling/13240
+//
+// m is number of blob, samples is a list of chunks
+// Inside the code, ft stands for first term; st for the second term; tt for the third term
+func (group *KzgEncoderGroup) UniversalVerify(params rs.EncodingParams, samples []Sample, m int) error {
+	// precheck
+	for i, s := range samples {
+		if s.Row >= m {
+			fmt.Printf("sample %v has %v Row, but there are only %v blobs\n", i, s.Row, m)
+			return errors.New("sample.Row and numBlob is inconsistend")
+		}
+	}
+
+	verifier, _ := group.GetKzgVerifier(params)
+	ks := verifier.Ks
+
+	D := params.ChunkLen
+
+	n := len(samples)
+	fmt.Printf("Batch verify %v frames of %v symbols out of %v blobs \n", n, params.ChunkLen, m)
+
+	r, err := GenRandomness(params, samples, m)
+	if err != nil {
+		return err
+	}
+
+	randomsFr := make([]bls.Fr, n)
+	bls.CopyFr(&randomsFr[0], &r)
+
+	// lhs
+	var tmp bls.Fr
+
+	// power of r
+	for j := 0; j < n-1; j++ {
+		bls.MulModFr(&randomsFr[j+1], &randomsFr[j], &r)
+	}
+
+	// array of proofs
+	proofs := make([]bls.G1Point, n)
+	for i := 0; i < n; i++ {
+		bls.CopyG1(&proofs[i], &samples[i].Proof)
+	}
+
+	// lhs g1
+	lhsG1 := bls.LinCombG1(proofs, randomsFr)
+
+	// lhs g2
+	lhsG2 := &ks.Srs.G2[D]
+
+	// rhs g2
+	rhsG2 := &bls.GenG2
+
+	// rhs g1
+	// get commitments
+	commits := make([]bls.G1Point, m)
+	// get coeffs
+	ftCoeffs := make([]bls.Fr, m)
+	for k := 0; k < n; k++ {
+		s := samples[k]
+		row := s.Row
+		bls.AddModFr(&ftCoeffs[row], &ftCoeffs[row], &randomsFr[k])
+		bls.CopyG1(&commits[row], &s.Commitment)
+	}
+
+	ftG1 := bls.LinCombG1(commits, ftCoeffs)
```

Comment:
It'll be more readability to name ft, st ... more properly. One option I see is: combinedProof, combinedCommitment, combinedInterpolation etc.

Reply:
I feel like it would be nice to find a way to show the hierarchy among the sections. It's unclear without prior knowledge that `//second term` refers to `//rhs g1`, for instance

Reply:
One thing could improve the readability is to break some parts into functions. I will rewrite and decompose some logics.

Reply:
@jianoaix @mooselumph I refactored the code, please take a look

Reply:
I would l
![Screenshot 2023-12-11 at 8 47 50 PM](https://github.com/Layr-Labs/eigenda/assets/99709935/e54b9e04-0c7e-49ac-9645-72009d7acca1)
ike this types of documentation in code, which makes the connection between code and theory clear:


I remember one of the highly useful documentation for me is to comment out that the coefficients in chunk is for interpolation polynomial, which connects the whole thing.